README.rst
=============
kotti_project
=============

Kotti_ is a high-level, Pythonic web application framework based on Pyramid_ and SQLAlchemy_.
It includes an extensible Content Management System called the Kotti CMS (see below).

Kotti is most useful when you are developing applications that

- have complex security requirements,
- use workflows, and/or
- work with hierarchical data.

Built on top of a number of *best-of-breed* software components,
most notably Pyramid_ and SQLAlchemy_,
Kotti introduces only a few concepts of its own,
thus hopefully keeping the learning curve flat for the developer.

.. _Kotti: http://kotti.pylonsproject.org/projects/pyramid/dev/
.. _Pyramid: http://docs.pylonsproject.org/projects/pyramid/dev/
.. _SQLAlchemy: http://www.sqlalchemy.org/

What is ``kotti_project``?
==========================
You can consider ``kotti_project`` as a meta-package or a Kotti CMS **distribution** with
the following goals:

* introduces the ``project`` concept
    * Enances the ``.ini``  and ``.txt`` setup for a structured approach suitable both for developing, 
      production or tag-based deployments. Without introducing a more complex layer like ``buildout``
* tries to be more friendly to non-Python folks who want to try out Kotti.
    * It might be difficult for non-Python developers understand things like install a virtualenv,
      activate the virtualenv and so on. That's why there is a regular ``Makefile`` that should
      make the life easier if you want to install a Kotti project and its prerequisites.

      Also ``Vagrant`` + ``Ansible`` provisioning available for an easy evaluation or automatic
      remote deployements
* third party plugins
    * includes the most useful third party plugins not included in the Kotti core,
      useful if you want to try out Kotti in a quick way
        * navigation (``kotti_navigation``)
        * events (``kotti_calendar``)
        * news (``kotti_news``)
    * you can also add other Kotti plugins published on PyPI. See https://pypi.python.org/pypi?%3Aaction=search&term=kotti&submit=search
* Kotti CMS "batteries included" installation
    * make commands for build and installation
    * right env setup
    * setup logging with logrotate enabled for production environments
    * working deploy examples (nginx, uwsgi, supervisor, database setup, etc). See the provisioning ``provisioning`` folder
      or run ``vagrant up``
    * sqlalchemy string connection examples for the most used relational databases
* Windows support
    * there is a module ready to be installed as a Windows service with a file system error log handler
* scaffolding/seed tool
    * you can generate your own custom projects starting from ``kotti_project`` (eg: ``$ kotti_project_clone kotti_anotherproject``)
* project **documentation** support
    * keep track of changes and write your project documentation with Sphinx, ready to be used (http://localhost:5000/docs/, requires admin privileges)
* easy assets customization
    * ``kotti-overrides`` ready to be used (override Kotti's templates and resources)
* deploy examples
    * provides working and ready to be used real world database connection examples and deploy methods (see ``provisioning``)

Installation
============

kotti_project installation
--------------------------

For development mode::

    $ make install-prerequisites    # it performs some ``apt-get install`` commands
    $ make install-dev
    $ make run-dev

Obviously if you are going to use a non-SQLite database you'll have to install other system packages and the
needed Python driver library.

If you need more info have a look at the official doc or if you need an example you can browse the ``provisioning``
folder of bootstrap the project with ``Vagrant``.

See::

    $ make help

for further options.

Vagrant + Ansible installation
==============================

You can use the provided Vagrant + Ansible setup:

1. checkout kotti_project

2. type ``vagrant up``

3. have a coffee

At the end of the process your Kotti instance will be served here http://localhost:5000. Login with admin / qwerty.


CHANGES.rst
CHANGELOG
=========

0.0.1 (unreleased)
------------------

- First release
  [davidemoro]



kottisvc_staging.py
# uncomment the next import line to get print to show up or see early
# exceptions if there are errors then run
#   python -m win32traceutil
# to see the output
# import win32traceutil
import win32serviceutil

PORT_TO_BIND = 8080
CONFIG_FILE = 'production_staging.ini'
SERVER_NAME = 'localhost'

SERVICE_NAME = "kotti_project_staging"
SERVICE_DISPLAY_NAME = "kotti_project_staging"
SERVICE_DESCRIPTION = """This will be displayed as a description \
of the serivice in the Services snap-in for the Microsoft \
Management Console."""


class PyWebService(win32serviceutil.ServiceFramework):
    """Python Web Service."""

    _svc_name_ = SERVICE_NAME
    _svc_display_name_ = SERVICE_DISPLAY_NAME
    _svc_deps_ = None        # sequence of service names on which this depends
    # Only exists on Windows 2000 or later, ignored on windows NT
    _svc_description_ = SERVICE_DESCRIPTION

    def SvcDoRun(self):
        import cherrypy
        from cherrypy import wsgiserver
        from pyramid.paster import get_app
        from pyramid.paster import setup_logging
        import os

        path = os.path.dirname(os.path.abspath(__file__))

        os.chdir(path)

        app = get_app(CONFIG_FILE)
        setup_logging(CONFIG_FILE)

        # See http://tools.cherrypy.org/wiki/WindowsService
        # See http://stackoverflow.com/questions/11982438/cherrypy-server-errors-log
        # If you have to debug cherrypy specific errors
        # enable the following option:
        #    'log.error_file': os.path.join(path, 'cherrypy_error.log'),
        cherrypy.config.update({
            'engine.autoreload.on': False,
            })
        self.server = wsgiserver.CherryPyWSGIServer(
            ('127.0.0.1', PORT_TO_BIND), app,
            numthreads=30,
            server_name=SERVER_NAME)

        self.server.start()

    def SvcStop(self):
        self.server.stop()


if __name__ == '__main__':
    win32serviceutil.HandleCommandLine(PyWebService)


kotti_project/scripts.py
import sys
import os
import string
import shutil
import optparse
import textwrap
import pkg_resources


def kotti_project_clone():
    description = """\
    Print the deployment settings for a Pyramid application.  Example:
    'show_settings deployment.ini'
    """
    usage = "usage: kotti_project_clone new_package_name"
    parser = optparse.OptionParser(
        usage=usage,
        description=textwrap.dedent(description)
        )

    options, args = parser.parse_args(sys.argv[1:])
    if not len(args) == 1:
        print('You must provide at least one argument')
        return 2

    print('Start translating package...')
    _translate(args[0])
    print('Done!')


def _translate(new_package_name, original_package_name='kotti_project'):
    """ Translate package """

    def package_split(package_name, separator='.'):
        """ With 'pippo.pluto.paperino' you'll get:
            ['pippo.pluto.paperino','pippo.pluto', 'pippo', 'pluto',
             'paperino']
        """
        items = package_name.split('.')
        len_items = len(items)
        results = [package_name, items[-1].capitalize()]

        for count in range(1, len_items-1):
            el = items[:-count]
            if el:
                results.append('.'.join(el))
        results.extend(items)
        return results

    # copy tree
    original_location = pkg_resources.\
        get_distribution(original_package_name).location
    shutil.copytree(
        original_location,
        new_package_name,
        ignore=shutil.ignore_patterns(
            '*.pyc', '.svn', '*.tmp', '*.egg-info', '*.egg',
            '.git', 'node_modules', '*.swp', '*~',))

    original_slices = package_split(original_package_name)
    new_slices = package_split(new_package_name)
    trans_map = zip(original_slices, new_slices)
    for step in os.walk(top=new_package_name, topdown=False):
        item_path = step[0]
        item_dir_paths = step[1]
        item_file_paths = step[2]

        for item_file_path in item_file_paths:
            # read the file and substitute patterns
            file_contents = ''
            # TODO: avoid rewriting of '*.png', '*.gif',
            # '*.svg', '*.ico', '*.jpg'
            # TODO: avoid duplicated open files
            with open(os.path.join(item_path,
                                   item_file_path), 'r') as file_to_filter:
                file_contents = file_to_filter.read()

            # translate file
            for trans in trans_map:
                old = trans[0]
                new = trans[1]
                file_contents = string.replace(file_contents, old, new)
                file_contents = string.replace(
                    file_contents,
                    old.capitalize(),
                    new.capitalize())

            # write translated version of file
            with open(os.path.join(item_path,
                                   item_file_path), 'w') as file_to_filter:
                file_to_filter.write(file_contents)

            # rename files
            for trans in trans_map:
                old = trans[0]
                new = trans[1]
                if old in item_file_path:
                    old_path = item_file_path
                    new_path = string.replace(old_path, old, new)
                    os.rename(os.path.join(item_path, old_path),
                              os.path.join(item_path, new_path))
                    break

        for item_dir_path in item_dir_paths:
            for trans in trans_map:
                old = trans[0]
                new = trans[1]
                if old in item_dir_path:
                    old_path = item_dir_path
                    new_path = string.replace(old_path, old, new)
                    os.rename(os.path.join(item_path, old_path),
                              os.path.join(item_path, new_path))
                    break


kottisvc.py
# uncomment the next import line to get print to show up or see early
# exceptions if there are errors then run
#   python -m win32traceutil
# to see the output
# import win32traceutil
import win32serviceutil

PORT_TO_BIND = 8080
CONFIG_FILE = 'production.ini'
SERVER_NAME = 'localhost'

SERVICE_NAME = "kotti_project"
SERVICE_DISPLAY_NAME = "kotti_project"
SERVICE_DESCRIPTION = """This will be displayed as a description \
of the serivice in the Services snap-in for the Microsoft \
Management Console."""


class PyWebService(win32serviceutil.ServiceFramework):
    """Python Web Service."""

    _svc_name_ = SERVICE_NAME
    _svc_display_name_ = SERVICE_DISPLAY_NAME
    _svc_deps_ = None        # sequence of service names on which this depends
    # Only exists on Windows 2000 or later, ignored on windows NT
    _svc_description_ = SERVICE_DESCRIPTION

    def SvcDoRun(self):
        import cherrypy
        from cherrypy import wsgiserver
        from pyramid.paster import get_app
        from pyramid.paster import setup_logging
        import os

        path = os.path.dirname(os.path.abspath(__file__))

        os.chdir(path)

        app = get_app(CONFIG_FILE)
        setup_logging(CONFIG_FILE)

        # See http://tools.cherrypy.org/wiki/WindowsService
        # See http://stackoverflow.com/questions/11982438/cherrypy-server-errors-log
        # If you have to debug cherrypy specific errors
        # enable the following option:
        #    'log.error_file': os.path.join(path, 'cherrypy_error.log'),
        cherrypy.config.update({
            'engine.autoreload.on': False,
            })
        self.server = wsgiserver.CherryPyWSGIServer(
            ('127.0.0.1', PORT_TO_BIND), app,
            numthreads=30,
            server_name=SERVER_NAME)

        self.server.start()

    def SvcStop(self):
        self.server.stop()


if __name__ == '__main__':
    win32serviceutil.HandleCommandLine(PyWebService)
